### ISSUE TITLE
Nice demo! / Ctrl+Z not working / Exit code not forwarded

### ISSUE THREAD

**Issue by hartwork** [shabbywu/simple-pty#1]
Hi @shabbywu,

very nice demo! I tried…

- `python3 -c 'from spty import spawn; import sys; spawn(sys.argv[1:])' htop`
- `python3 -c 'from spty import spawn; import sys; spawn(sys.argv[1:])' sh -c 'exit 123' ; echo $?`

…and learned that…

- <kbd>Ctrl</kbd>+<kbd>Z</kbd> is not working in htop and that
- the exit code is not passed through in the second example.

Would you be interested in fixing these issues?

**Comment by shabbywu** [comment]
Thanks your report, I will have a try to fix it.


But i will have a long time holiday plan this moon so i am afraid to will not make a promise to fix it as soon as possible.

**Comment by hartwork** [comment]
@shabbywu thanks for your quick reply! 

While I'm still figuring things out about it, I would like to share that one piece needed to make this work seems to be re-enabling "isig" mode with the pty to even be able to receive the signal, e.g.:

```python
    import subprocess
    subprocess.check_call(['stty', 'isig'])

    def handle_sigtstp(*args, **kwargs):
        print('got SIGTSTP')

    signal.signal(signal.SIGTSTP, handle_sigtstp)
```

Sending `os.kill(pid, signal.SIGSTOP)` inside the handler seems to be part of the solution, but it needs more, not yet sure what exactly. Maybe you already know.

Have a nice holiday!

**Comment by shabbywu** [comment]
Yeah, add more signal lisener will work well. For a better perfection, I may need to refer to the implementation of bash or [zsh](https://github.com/zsh-users/zsh/blob/9f57ca4ac8ae071727b1d77cbb8c4c0d893b9099/Src/signals.c#L593)

**Comment by shabbywu** [comment]
@hartwork Have a try.

```bash
❯ pip install https://codeload.github.com/shabbywu/simple-pty/zip/refs/heads/issues-001

# now spawn will return the exitcode of subprocess
❯ python3 -c 'from spty import spawn; import sys; sys.exit(spawn(sys.argv[1:])[1])' sh -c 'exit 123' ; echo $?

# now spawn will return the resumer when accept SIGTSTP
❯ python3 -c 'from spty import spawn; import sys; resume = spawn(sys.argv[1:])[2]; print("hang up for 20s!!!"); import time; time.sleep(20); print("resume!!!"); resume()' bash
```

**Comment by hartwork** [comment]
@shabbywu that was quick!

I played with it a but it's not there yet, or I'm doing something wrong:
- When I press Ctrl+C while running cat, the return code I'd expected is 128+2 = 130 while simple-pty returns 2.
- When I press Ctrl+Z from inside htop:
  - The expected behavior would be that the process receives `SIGSTOP`, the htop "window" disappears, Bash says something like `[1]+ Stopped htop`, and running `fg` will continue htop execution and bring the window back up.
  - The actual behavior is that (on pressing Ctrl+Z in htop behind simple-pty) the htop window "freezes" but remains shown, I get a new Bash prompt right beneath it, running `fg` gets me `bash: fg: current: no such job`, and the htop process is gone.

What do you think?

**Comment by shabbywu** [comment]
## For the first question

Actually, exitcode of the process killed by signal SIGINT is 2(is also **the signal number of SIGINT**), but some kind of shell such as zsh, bash, sh will bonus it to 130, where is 128 + 2.
And this is not a ***de facto norm***, for there have another shell will bonus signal number with 256 or 384(256+128).
> Reference: https://unix.stackexchange.com/a/99134


So let us come back to this issue. As you said, the processes interrupted by Ctrl+C will return the exitcode 2. 

**And Why?**

Just because **simple-pty is a pseudo terminal but not a shell**. I thinks simple-pty should and must return the exitcode as it is. 
But I also thinks if simple-pty can provided some function for programmer to get the latest signal or can provided a function to convert exticode to bash-like exitcode will be cool too.

**Comment by shabbywu** [comment]
## For the second question

As we known, **fg** is a builtin magic behavior for shell, And i have implemented a similar action to resume background process in simple-pty@v2.0.0. 
You can resume the background process to frontground by calling the `resume` callback returned from `spawn`. For example:

```python
Python 3.8.13 (default, Aug  3 2022, 15:09:30)
[Clang 13.1.6 (clang-1316.0.21.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from spty import spawn
>>> pid, exitcode, resume = spawn("./htop")   # after call it, htop will be the frontground process instead of python. press `Ctrl+Z` to put it back to background.
>>> print("After press `Ctrl+Z`, the control is back to python interactive console")
>>> resume()  # call `resume` to put `htop` to frontground again, just like `fg %1`
```


Yeah, this `resume` api is not pythonic, but like something work in javascript. What do you thinks how the api should be? 

For example:
```python
>>> from spty import fg, spawn
>>> pid, exitcode = spawn("./htop")  # press `Ctrl+Z`
>>> fg(pid)  # resume it frontground 
```

---
By the way, the htop window "freezes" is a bug...
According to my observation, something should be restored after press `Ctrl+Z`(just like press `Ctrl+C`), but now i dont know what should it be restored.

**Comment by shabbywu** [comment]
Hey, I have not fully test simple-pty@2.0.0, it should be have some bug.
The branch issue-001 is only a preview version. 

If you have some suggestion of the new api or bug report, please let me know. Thanks

**Comment by hartwork** [comment]
@shabbywu I think we should separate use of API from behavior. Let's first get the behavior right, and then potentially API use just needs an additional wrapper around `spawn` and resuming, maybe.

Regarding exit code and signals, if those two channels would be served separately, that would fix the whole problem: On integer for a "true" exit code, one integer for a "true" signal code. 

I'm not sure yet how the new code would be used. It would need a loop, that much is clear, but I seem to have some misunderstanding. This is my current play code:

```python
import signal
import sys
import termios

from spty import spawn

old_state = termios.tcgetattr(sys.stdin.fileno())

_child_pid, exit_code_or_signal, resume_child = spawn(['htop'])
while exit_code_or_signal == -1:
    if exit_code_or_signal == signal.SIGTSTP:
        new_state = termios.tcgetattr(sys.stdin.fileno())
        termios.tcsetattr(sys.stdin.fileno(), termios.TCSAFLUSH, old_state)

        os.kill(0, signal.SIGSTOP)

        termios.tcsetattr(sys.stdin.fileno(), termios.TCSAFLUSH, new_state)

    assert resume_child is not None
    _child_pid, exit_code_or_signal, resume_child = resume_child()

termios.tcsetattr(sys.stdin.fileno(), termios.TCSAFLUSH, old_state)

sys.exit(exit_code_or_signal)

```

The symptom is that that first Ctrl+Z seems ignored, and then the second kills both processes altogether.
Maybe I got the idea wrong. Can you take that snippet and make feel like we'd be talking to unwrapped `htop`? Once we have that behavior, the rest will be easy, I think.

**Comment by shabbywu** [comment]
You are right, i thinks we should take TDD. After my holiday, I will take some test case to insure the behavior right, and then make the API readable.

**Comment by shabbywu** [comment]
I have update the branch [`issues-001`](https://github.com/shabbywu/simple-pty/tree/issues-001)
- fix bug: signal handler will not restore
- refactor the API, for better usage and readable
- add a test case(Also an example) for running [top](https://github.com/shabbywu/simple-pty/blob/issues-001/integration/test-top/test.py) with simple-pty

---

And the thing I must clarify is that `simple-pty` have done reset tty attributes too.
So you do not need to cache and restore tty attributes.

---
And now we can run `htop` by the code below:
```python
#!/usr/bin/env python
import sys
import time

from spty import spawn, fg  # new api


child_pid, exitcode = spawn(["top"])
# Note: Maybe there should be more information for why the process is hung?
# -1 for the process have not exited, 
while exitcode == -1:
    # the process is suspended by `Ctrl + Z`
    # resume if ten seconds after
    time.sleep(10)
    child_pid, exitcode = fg(child_pid)

# the process is exited, return the exitcode as it is
sys.exit(exitcode)
```

A screenshot is that, but still have some bugs:
- when press `Ctrl +Z`. The terminal will not restore for some unknown trouble. 
- And after resumed the terminal will not refresh too.

![image](https://github.com/shabbywu/simple-pty/assets/5237578/de01d16c-8514-4405-9712-a3acad9094d1)

victory is in sight, i think.

**Comment by hartwork** [comment]
@shabbywu there is a C implementation aiming for the same at https://github.com/Duncaen/runpty/blob/main/runpty.c and it does things like calling `setsid` and also adjusting process groups and a lot more. Maybe we need more of these things also, I don't understand them in detail yet.
